### PR TITLE
Updated workflows: Added separate jobs and Docker Hub push

### PR DESCRIPTION
**Pull request by AhmedHadzihamdo**
### Changes:
- Added separate jobs for backend and frontend in both Pull Request and Merge workflows.
- Included caching for dependencies in Pull Request workflow.
- Enabled pushing Docker images to Docker Hub in Merge workflow.

### Notes:
- Please review and approve before merging.

### CODE REVIEW

File: .github/workflows/merge.yml [hunk 1]
```diff
@@ -1,34 +1,43 @@
...
-          node-version: '20'
+      - name: Build Backend Docker Image
+        working-directory: ./backend
+        run: |
+          docker build -t petclinic-backend:latest .
+          docker tag petclinic-backend:latest ahmedhamdo/petclinic-backend:latest
+
+      - name: Push Backend Docker Image to Docker Hub
+        run: |
+          echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u "${{ secrets.DOCKER_USERNAME }}" --password-stdin
```
Comment (mujohadzimehanovic-abh) [inline .github/workflows/merge.yml]: There is a github action for Docker login, lets use it.
https://github.com/docker/login-action

File: .github/workflows/merge.yml [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
-        uses: actions/setup-node@v3
-        with:
-          node-version: '20'
+      - name: Build Backend Docker Image
+        working-directory: ./backend
+        run: |
+          docker build -t petclinic-backend:latest .
+          docker tag petclinic-backend:latest ahmedhamdo/petclinic-backend:latest
+
+      - name: Push Backend Docker Image to Docker Hub
```
Comment (mujohadzimehanovic-abh) [inline .github/workflows/merge.yml]: There is a github action for build&push.
https://github.com/docker/build-push-action

File: .github/workflows/merge.yml [hunk 3]
```diff
@@ -1,34 +1,43 @@
...
+          echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u "${{ secrets.DOCKER_USERNAME }}" --password-stdin
+          docker push ahmedhamdo/petclinic-backend:latest
+
+  build-frontend:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
 
       - name: Build Frontend Docker Image
```
Comment (mujohadzimehanovic-abh) [inline .github/workflows/merge.yml]: Same thing as previous comments, lets use github action.

File: .github/workflows/pull_request.yml [hunk 4]
```diff
@@ -33,3 +29,28 @@ jobs:
       - name: Build Backend
         working-directory: ./backend
         run: mvn clean install
```
Comment (mujohadzimehanovic-abh) [inline .github/workflows/pull_request.yml]: Would it make sense to run tests here ?

File: .github/workflows/pull_request.yml [hunk 5]
```diff
@@ -33,3 +29,28 @@ jobs:
...
+          key: npm-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            npm-
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: '20'
+
+      - name: Install and Build Frontend
```
Comment (mujohadzimehanovic-abh) [inline .github/workflows/pull_request.yml]: Here the same question comment for tests ?

**Review by mujohadzimehanovic-abh — CHANGES_REQUESTED** [review]

**Comment by AhmedHadzihamdo** [comment]
I made changes to merge.yml. Can you review this?

**Comment by mujohadzimehanovic-abh** [comment]
Tests comments were for pull_request job.

**Comment by AhmedHadzihamdo** [comment]
I added frontend tests to pull_request.yml and removed tests from merge.yml, please review changes.

**Pull request merged**